File: packages/typescript/goldenmatch/src/core/indicators.ts
/**
 * indicators.ts — Auto-config complexity indicators (v1.10).
 *
 * TS port of ``packages/python/goldenmatch/goldenmatch/core/indicators.py``.
 * Pure functions: each takes a row array (and optional config args) and
 * returns a typed result. No controller state, no I/O. Each function has
 * a wall-clock budget; on exhaustion, returns null or a sentinel.
 *
 * The Python implementation operates on polars DataFrames; we mirror the
 * algorithms on plain row-of-object arrays so the module stays edge-safe
 * (no ``node:`` imports, no polars).
 */

import type { Row, GoldenMatchConfig } from "./types.js";
import { getMatchkeys } from "./types.js";
import type {
  ColumnPrior,
  SparsityVerdict,
  CollisionSignal,
} from "./complexityProfile.js";

// ---------------------------------------------------------------------------
// Wall-clock budgets (seconds) — match Python constants.
// ---------------------------------------------------------------------------

export const BUDGET_COLUMN_PRIORS = 5.0;
export const BUDGET_SPARSE_MATCH = 2.0;
export const BUDGET_FULL_POP_HITS = 15.0;
export const BUDGET_CROSS_BLOCKING = 20.0;
export const BUDGET_CORRUPTION = 3.0;
export const BUDGET_COLLISION = 8.0;

// Identity-column name → identity_score floor heuristics. Mirrors
// ``_IDENTITY_NAME_PATTERNS`` in Python.
const IDENTITY_NAME_PATTERNS: ReadonlyArray<readonly [RegExp, number]> = [
  [/^(email|e[-_]?mail|email_addr)$/i, 0.95],
  [/^(ssn|social|tax_id)$/i, 0.95],
  [/^(phone|mobile|tel|telephone)$/i, 0.85],
  [/^(id|uuid|guid|user_id|account_id)$/i, 0.9],
];

function elapsedSec(startMs: number): number {
  return (Date.now() - startMs) / 1000;
}

function userColumns(rows: readonly Row[]): string[] {
  if (rows.length === 0) return [];
  return Object.keys(rows[0] as object);
}

// Lightweight column dtype probe: returns "boolean" / "date" / "other".
// Python's `_NON_IDENTITY_DTYPES` returns identity_score 0 for booleans/dates.
function columnKind(rows: readonly Row[], col: string): "boolean" | "date" | "other" {
  for (const r of rows) {
    const v = (r as Record<string, unknown>)[col];
    if (v === null || v === undefined || v === "") continue;
    if (typeof v === "boolean") return "boolean";
    if (v instanceof Date) return "date";
    return "other";
  }
  return "other";
}

// ---------------------------------------------------------------------------
// Indicator 1: compute_column_priors
// ---------------------------------------------------------------------------

/** Per-column identity + corruption priors. Returns an empty record for empty data. */
export function computeColumnPriors(
  rows: readonly Row[],
): Record<string, ColumnPrior> {
  const start = Date.now();
  if (rows.length === 0) return {};
  const cols = userColumns(rows);
  const sample = rows.length > 1000 ? rows.slice(0, 1000) : rows;
  const priors: Record<string, ColumnPrior> = {};

  for (const col of cols) {
    if (elapsedSec(start) > BUDGET_COLUMN_PRIORS) {
      for (const remaining of cols) {
        if (!(remaining in priors)) {
          priors[remaining] = { identityScore: 0.0, corruptionScore: 0.0 };
        }
      }
      break;
    }
    const identityScore = computeIdentityScore(rows, col);
    const corruptionScore = computeCorruptionScoreInline(sample, col);
    priors[col] = { identityScore, corruptionScore };
  }
  return priors;
}

function computeIdentityScore(rows: readonly Row[], col: string): number {
  const kind = columnKind(rows, col);
  if (kind === "boolean" || kind === "date") return 0.0;
  for (const [pat, score] of IDENTITY_NAME_PATTERNS) {
    if (pat.test(col)) return score;
  }
  // High-cardinality column = id-like. Python uses polars' ``n_unique``
  // which counts nulls as one distinct value too — mirror that here so the
  // ratio matches for all-null / mostly-null columns.
  const distinct = new Set<unknown>();
  for (const r of rows) {
    const v = (r as Record<string, unknown>)[col];
    // Treat null/undefined/empty-string as a single "null" sentinel.
    distinct.add(v === null || v === undefined || v === "" ? "__NULL__" : v);
  }
  const total = rows.length;
  const cardinalityRatio = distinct.size / Math.max(1, total);
  if (cardinalityRatio > 0.5) return 0.7;
  if (cardinalityRatio > 0.1) return 0.3;
  return 0.0;
}

function computeCorruptionScoreInline(
  sample: readonly Row[],
  col: string,
): number {
  if (BUDGET_CORRUPTION <= 0.0) return 0.0;
  const vals: string[] = [];
  for (const r of sample) {
    const v = (r as Record<string, unknown>)[col];
    if (v === null || v === undefined) {
      vals.push("");
      continue;
    }
    vals.push(String(v));
  }
  if (vals.length === 0) return 0.0;
  const raw = new Set<string>();
  const normalized = new Set<string>();
  for (const v of vals) {
    if (!v) continue;
    raw.add(v);
    normalized.add(v.trim().toLowerCase());
  }
  if (raw.size === 0) return 0.0;
  const ratioClean = normalized.size / raw.size;
  return Math.max(0.0, Math.min(1.0, 1.0 - ratioClean));
}

/** Public per-column corruption score (mirrors ``compute_corruption_score``). */
export function computeCorruptionScore(rows: readonly Row[], col: string): number {
  if (rows.length === 0) return 0.0;
  const cols = userColumns(rows);
  if (!cols.includes(col)) return 0.0;
  const sample = rows.length > 1000 ? rows.slice(0, 1000) : rows;
  return computeCorruptionScoreInline(sample, col);
}

// ---------------------------------------------------------------------------
// Indicator 2: estimate_sparse_match_signal
// ---------------------------------------------------------------------------

export interface SparseMatchOptions {
  readonly exactColumns?: readonly string[];
  readonly sampleSize?: number;
  readonly sparseThreshold?: number;
}

export function estimateSparseMatchSignal(
  rows: readonly Row[],
  options: SparseMatchOptions = {},
): SparsityVerdict {
  const exactColumns = options.exactColumns ?? [];
  const sampleSize = options.sampleSize ?? 1000;
  const sparseThreshold = options.sparseThreshold ?? 50;
  if (exactColumns.length === 0 || rows.length === 0) {
    return { isSparse: true, estimatedNTruePairs: 0 };
  }
  const sample = rows.length > sampleSize ? rows.slice(0, sampleSize) : rows;
  const cols = userColumns(sample);
  const NULL_KEY = "__NULL__";
  let nPairs = 0;
  for (const col of exactColumns) {
    if (!cols.includes(col)) continue;
    const counts = new Map<unknown, number>();
    for (const r of sample) {
      const v = (r as Record<string, unknown>)[col];
      const key = v === null || v === undefined || v === "" ? NULL_KEY : v;
      counts.set(key, (counts.get(key) ?? 0) + 1);
    }
    for (const n of counts.values()) {
      if (n > 1) nPairs += (n * (n - 1)) / 2;
    }
  }
  return { isSparse: nPairs < sparseThreshold, estimatedNTruePairs: nPairs };
}

// ---------------------------------------------------------------------------
// Indicator 3: estimate_full_pop_hits
// ---------------------------------------------------------------------------

export function estimateFullPopHits(
  rows: readonly Row[],
  blockingCol: string,
): number | null {
  if (BUDGET_FULL_POP_HITS <= 0.0) return null;
  const start = Date.now();
  if (rows.length === 0) return 0;
  const cols = userColumns(rows);
  if (!cols.includes(blockingCol)) return 0;
  if (elapsedSec(start) > BUDGET_FULL_POP_HITS) return null;
  try {
    // Python parity: polars ``group_by`` keeps nulls as a single group, so
    // multiple null rows in the blocking column count as a collision group.
    const counts = new Map<unknown, number>();
    const NULL_KEY = "__NULL__";
    for (const r of rows) {
      const raw = (r as Record<string, unknown>)[blockingCol];
      const key = raw === null || raw === undefined || raw === "" ? NULL_KEY : raw;
      counts.set(key, (counts.get(key) ?? 0) + 1);
    }
    if (elapsedSec(start) > BUDGET_FULL_POP_HITS) return null;
    let nPairs = 0;
    for (const n of counts.values()) {
      if (n > 1) nPairs += (n * (n - 1)) / 2;
    }
    return nPairs;
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Indicator 4: compute_cross_blocking_overlap
// ---------------------------------------------------------------------------

export function computeCrossBlockingOverlap(
  rows: readonly Row[],
  keyA: string,
  keyB: string,
): number | null {
  if (keyA === keyB) return 1.0;
  if (BUDGET_CROSS_BLOCKING <= 0.0) return null;
  const start = Date.now();
  const cols = userColumns(rows);
  if (!cols.includes(keyA) || !cols.includes(keyB) || rows.length === 0) {
    return null;
  }
  if (elapsedSec(start) > BUDGET_CROSS_BLOCKING) return null;
  try {
    // Python parity: polars ``group_by`` includes nulls as a single group.
    const NULL_KEY = "__NULL__";
    const groupA = new Map<unknown, number[]>();
    const groupB = new Map<unknown, number[]>();
    for (let i = 0; i < rows.length; i++) {
      const r = rows[i] as Record<string, unknown>;
      const va = r[keyA];
      const vb = r[keyB];
      const ka = va === null || va === undefined || va === "" ? NULL_KEY : va;
      const kb = vb === null || vb === undefined || vb === "" ? NULL_KEY : vb;
      {
        const arr = groupA.get(ka);
        if (arr) arr.push(i);
        else groupA.set(ka, [i]);
      }
      {
        const arr = groupB.get(kb);
        if (arr) arr.push(i);
        else groupB.set(kb, [i]);
      }
    }
    if (elapsedSec(start) > BUDGET_CROSS_BLOCKING) return null;
    const pairsSet = (grouped: Map<unknown, number[]>): Set<string> | null => {
      const pairs = new Set<string>();
      for (const arr of grouped.values()) {
        if (arr.length < 2) continue;
        const sorted = arr.slice().sort((a, b) => a - b);
        for (let i = 0; i < sorted.length; i++) {
          for (let j = i + 1; j < sorted.length; j++) {
            pairs.add(`${sorted[i]}|${sorted[j]}`);
            if (elapsedSec(start) > BUDGET_CROSS_BLOCKING) return null;
          }
        }
      }
      return pairs;
    };
    const setA = pairsSet(groupA);
    if (setA === null) return null;
    const setB = pairsSet(groupB);
    if (setB === null) return null;
    const union = new Set<string>(setA);
    for (const p of setB) union.add(p);
    if (union.size === 0) return 1.0;
    let inter = 0;
    for (const p of setA) if (setB.has(p)) inter += 1;
    return inter / union.size;
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Indicator 5: compute_identity_collision_signal
// ---------------------------------------------------------------------------

/**
 * Detect whether an identity column is shared across distinct entities.
 *
 * For each multi-record group, compute the max pairwise divergence
 * (1 - similarity) on witnesses, using a token-sort ratio approximation.
 * Returns the fraction of multi-record groups with max-divergence > 0.5.
 */
export function computeIdentityCollisionSignal(
  rows: readonly Row[],
  identityCol: string,
  witnessCols: readonly string[],
): CollisionSignal {
  const start = Date.now();
  if (BUDGET_COLLISION <= 0.0) return { rate: 0.0, witnessUsed: "" };
  if (witnessCols.length === 0 || rows.length === 0) {
    return { rate: 0.0, witnessUsed: "" };
  }
  const cols = userColumns(rows);
  if (!cols.includes(identityCol)) return { rate: 0.0, witnessUsed: "" };
  const valid = witnessCols.filter((c) => cols.includes(c));
  if (valid.length === 0) return { rate: 0.0, witnessUsed: "" };

  try {
    const groups = new Map<unknown, number[]>();
    for (let i = 0; i < rows.length; i++) {
      const v = (rows[i] as Record<string, unknown>)[identityCol];
      if (v === null || v === undefined || v === "") continue;
      const arr = groups.get(v);
      if (arr) arr.push(i);
      else groups.set(v, [i]);
    }
    const multi: number[][] = [];
    for (const arr of groups.values()) if (arr.length > 1) multi.push(arr);
    if (elapsedSec(start) > BUDGET_COLLISION || multi.length === 0) {
      return { rate: 0.0, witnessUsed: "" };
    }

    let nGroups = multi.length;
    let nHighDivergence = 0;
    let winningWitness = "";
    let maxObservedDiv = 0.0;
    for (const groupRows of multi) {
      if (elapsedSec(start) > BUDGET_COLLISION) {
        return { rate: 0.0, witnessUsed: "" };
      }
      let maxDivInGroup = 0.0;
      for (const witness of valid) {
        const vals = groupRows.map((i) => {
          const v = (rows[i] as Record<string, unknown>)[witness];
          return v === null || v === undefined ? "" : String(v);
        });
        for (let i = 0; i < vals.length; i++) {
          for (let j = i + 1; j < vals.length; j++) {
            const sim = tokenSortRatio(vals[i]!, vals[j]!);
            const div = 1.0 - sim;
            if (div > maxDivInGroup) {
              maxDivInGroup = div;
              if (div > maxObservedDiv) {
                maxObservedDiv = div;
                winningWitness = witness;
              }
            }
          }
        }
      }
      if (maxDivInGroup > 0.5) nHighDivergence += 1;
    }
    const rate = nGroups > 0 ? nHighDivergence / nGroups : 0.0;
    return { rate, witnessUsed: winningWitness };
  } catch {
    return { rate: 0.0, witnessUsed: "" };
  }
}

/** Token-sort ratio approximation: sort whitespace-separated tokens, then
 *  compute a Levenshtein-derived similarity (0..1). */
function tokenSortRatio(a: string, b: string): number {
  const ta = a.toLowerCase().trim().split(/\s+/).filter(Boolean).sort().join(" ");
  const tb = b.toLowerCase().trim().split(/\s+/).filter(Boolean).sort().join(" ");
  if (ta === tb) return 1.0;
  if (ta.length === 0 && tb.length === 0) return 1.0;
  const dist = levenshtein(ta, tb);
  const maxLen = Math.max(ta.length, tb.length);
  return maxLen === 0 ? 1.0 : 1.0 - dist / maxLen;
}

function levenshtein(a: string, b: string): number {
  if (a === b) return 0;
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;
  const prev = new Array<number>(b.length + 1);
  const curr = new Array<number>(b.length + 1);
  for (let j = 0; j <= b.length; j++) prev[j] = j;
  for (let i = 1; i <= a.length; i++) {
    curr[0] = i;
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(curr[j - 1]! + 1, prev[j]! + 1, prev[j - 1]! + cost);
    }
    for (let j = 0; j <= b.length; j++) prev[j] = curr[j]!;
  }
  return prev[b.length]!;
}

// ---------------------------------------------------------------------------
// IndicatorContext — memoization layer (matches Python's class shape)
// ---------------------------------------------------------------------------

/**
 * Context object passed into indicator-aware refit rules. Memoizes each
 * indicator's result so repeated rule lookups in one iteration share work.
 *
 * Mirrors Python ``goldenmatch.core.autoconfig_controller.IndicatorContext``.
 */
export class IndicatorContext {
  readonly rows: readonly Row[];
  readonly config: GoldenMatchConfig;
  private _columnPriors: Record<string, ColumnPrior> | null = null;
  private _sparsity: SparsityVerdict | null = null;
  private _fullPop: Map<string, number | null> = new Map();
  private _crossOverlap: Map<string, number | null> = new Map();
  private _collision: Map<string, CollisionSignal> = new Map();
  private _firedFlags: Set<string> = new Set();

  constructor(rows: readonly Row[], config: GoldenMatchConfig) {
    this.rows = rows;
    this.config = config;
  }

  get columnPriors(): Record<string, ColumnPrior> {
    if (this._columnPriors === null) {
      this._columnPriors = computeColumnPriors(this.rows);
    }
    return this._columnPriors;
  }

  get sparsityVerdict(): SparsityVerdict {
    if (this._sparsity === null) {
      const exactCols: string[] = [];
      for (const mk of getMatchkeys(this.config)) {
        if (mk.type === "exact") {
          for (const f of mk.fields) exactCols.push(f.field);
        }
      }
      this._sparsity = estimateSparseMatchSignal(this.rows, {
        exactColumns: exactCols,
      });
    }
    return this._sparsity;
  }

  fullPopHits(blockingCol: string): number | null {
    if (!this._fullPop.has(blockingCol)) {
      this._fullPop.set(blockingCol, estimateFullPopHits(this.rows, blockingCol));
    }
    return this._fullPop.get(blockingCol) ?? null;
  }

  crossBlockingOverlap(keyA: string, keyB: string): number | null {
    const cacheKey = keyA <= keyB ? `${keyA}|${keyB}` : `${keyB}|${keyA}`;
    if (!this._crossOverlap.has(cacheKey)) {
      this._crossOverlap.set(
        cacheKey,
        computeCrossBlockingOverlap(this.rows, keyA, keyB),
      );
    }
    return this._crossOverlap.get(cacheKey) ?? null;
  }

  collisionSignal(identityCol: string, witnessCols: readonly string[]): CollisionSignal {
    const cacheKey = `${identityCol}|${witnessCols.slice().sort().join(",")}`;
    if (!this._collision.has(cacheKey)) {
      this._collision.set(
        cacheKey,
        computeIdentityCollisionSignal(this.rows, identityCol, witnessCols),
      );
    }
    return this._collision.get(cacheKey) ?? { rate: 0.0, witnessUsed: "" };
  }

  /** Side-channel flag used by ``rule_sparse_match_expand`` to one-shot. */
  hasFired(name: string): boolean {
    return this._firedFlags.has(name);
  }
  markFired(name: string): void {
    this._firedFlags.add(name);
  }

  /** Columns of the backing data (mirrors Python's ``ctx._df.columns``). */
  get columns(): readonly string[] {
    return userColumns(this.rows);
  }
}

File: packages/python/goldenmatch/goldenmatch/core/indicators.py
"""Auto-config complexity indicators (v1.10).

Pure functions: each takes a polars DataFrame (and optional config args)
and returns a typed result. No controller state, no I/O. Each function
has a wall-clock budget; on exhaustion, returns None or a sentinel.

Spec: docs/superpowers/specs/2026-05-08-autoconfig-indicators-design.md
"""
from __future__ import annotations

import logging
import re
import time

import polars as pl

from goldenmatch.core.complexity_profile import (
    CollisionSignal,
    ColumnPrior,
    SparsityVerdict,
)

logger = logging.getLogger(__name__)

# Wall-clock budgets (seconds). Indicators returning None on budget
# exhaustion are documented in the spec § Error handling.
BUDGET_COLUMN_PRIORS = 5.0
BUDGET_SPARSE_MATCH = 2.0
BUDGET_FULL_POP_HITS = 15.0
BUDGET_CROSS_BLOCKING = 20.0
BUDGET_CORRUPTION = 3.0

# Identity-column heuristics. Column-name regex -> identity_score floor.
# These are NOT authoritative -- they're priors. Real identity verification
# happens via full-pop matchkey hits.
_IDENTITY_NAME_PATTERNS = [
    (re.compile(r"^(email|e[-_]?mail|email_addr)$", re.I), 0.95),
    (re.compile(r"^(ssn|social|tax_id)$", re.I), 0.95),
    (re.compile(r"^(phone|mobile|tel|telephone)$", re.I), 0.85),
    (re.compile(r"^(id|uuid|guid|user_id|account_id)$", re.I), 0.90),
]

_BOOLEAN_DTYPES = {pl.Boolean}
_NON_IDENTITY_DTYPES = {pl.Boolean, pl.Date, pl.Datetime, pl.Time}


def compute_column_priors(df: pl.DataFrame) -> dict[str, ColumnPrior]:
    """Compute per-column identity + corruption priors.

    identity_score:
      - 0.95 for canonical identity column names (email/ssn/uuid)
      - 0.85 for phone-like
      - 0.7 for high-cardinality strings (cardinality_ratio > 0.5)
      - 0.0 for booleans/dates/categoricals/low-cardinality

    corruption_score:
      - Computed on a 1000-row sample
      - High when within-column edit-distance variance is high (Brian/BRIAN/B.)
      - Low when entries are deterministic (clean email patterns)
    """
    start = time.time()
    if df.is_empty():
        return {}
    priors: dict[str, ColumnPrior] = {}
    sample = df.head(1000) if df.height > 1000 else df

    for col in df.columns:
        if (time.time() - start) > BUDGET_COLUMN_PRIORS:
            logger.info(
                "compute_column_priors: budget %ss exceeded; "
                "remaining %d columns get default priors",
                BUDGET_COLUMN_PRIORS, len(df.columns) - len(priors),
            )
            for remaining in df.columns:
                if remaining not in priors:
                    priors[remaining] = ColumnPrior(0.0, 0.0)
            break

        identity_score = _compute_identity_score(df, col)
        corruption_score = _compute_corruption_score_inline(sample, col)
        priors[col] = ColumnPrior(
            identity_score=identity_score,
            corruption_score=corruption_score,
        )

    return priors


def _compute_identity_score(df: pl.DataFrame, col: str) -> float:
    """Identity-score heuristic. Name match > dtype match > cardinality."""
    if df.schema.get(col) in _NON_IDENTITY_DTYPES:
        return 0.0
    for pattern, score in _IDENTITY_NAME_PATTERNS:
        if pattern.match(col):
            return score
    # High-cardinality string column = id-like
    try:
        n_unique = df[col].n_unique()
        cardinality_ratio = n_unique / max(1, df.height)
        if cardinality_ratio > 0.5:
            return 0.7
        if cardinality_ratio > 0.1:
            return 0.3
    except Exception:
        pass
    return 0.0


def _compute_corruption_score_inline(sample: pl.DataFrame, col: str) -> float:
    """Approximation of within-column edit-distance variance.

    Cheap proxy: fraction of values that are case-or-whitespace-collapsed
    duplicates of another value in the sample. High value -> high
    corruption (Brian/BRIAN/brian/Brian ).
    """
    # Pre-flight budget check: BUDGET_CORRUPTION <= 0.0 means "disabled".
    if BUDGET_CORRUPTION <= 0.0:
        return 0.0
    start = time.time()
    try:
        vals = sample[col].cast(str).fill_null("").to_list()
    except Exception:
        return 0.0
    if (time.time() - start) > BUDGET_CORRUPTION:
        return 0.0
    if not vals:
        return 0.0
    normalized = {v.strip().lower() for v in vals if v}
    raw = {v for v in vals if v}
    if not raw:
        return 0.0
    # Corruption ratio: how many distinct raw forms collapse to fewer
    # normalized forms. 1.0 means perfectly clean (1:1); lower means noise.
    ratio_clean = len(normalized) / len(raw)
    # Invert and clamp: corruption_score = 1.0 - ratio_clean
    return max(0.0, min(1.0, 1.0 - ratio_clean))


def estimate_sparse_match_signal(
    df: pl.DataFrame,
    exact_columns: list[str] | None = None,
    sample_size: int = 1000,
    sparse_threshold: int = 50,
) -> SparsityVerdict:
    """Count exact-matchkey collisions in a sample.

    If `exact_columns` is empty (caller has no exact matchkeys), treat as
    sparse -- controller can't sanity-check otherwise.
    """
    if not exact_columns or df.is_empty():
        return SparsityVerdict(is_sparse=True, estimated_n_true_pairs=0)
    sample = df.head(sample_size) if df.height > sample_size else df
    n_pairs = 0
    for col in exact_columns:
        if col not in sample.columns:
            continue
        try:
            counts = (
                sample.group_by(col)
                .agg(pl.len().alias("n"))
                .filter(pl.col("n") > 1)
            )
            # Each group of size k contributes k*(k-1)/2 pairs
            n_pairs += int(
                counts.select(
                    (pl.col("n") * (pl.col("n") - 1) / 2).sum()
                ).item()
            )
        except Exception:
            continue
    is_sparse = n_pairs < sparse_threshold
    return SparsityVerdict(is_sparse=is_sparse, estimated_n_true_pairs=n_pairs)


def compute_corruption_score(df: pl.DataFrame, col: str) -> float:
    """Public API for per-column corruption score (case/whitespace noise).

    See _compute_corruption_score_inline for the heuristic.
    """
    if col not in df.columns or df.is_empty():
        return 0.0
    sample = df.head(1000) if df.height > 1000 else df
    return _compute_corruption_score_inline(sample, col)


def estimate_full_pop_hits(df: pl.DataFrame, blocking_col: str) -> int | None:
    """Count exact-match collisions on the full population.

    Returns None on budget exhaustion. Used by indicator-aware rules
    to validate that v0's blocking key has structural signal even when
    sample's mass_above_threshold == 0.
    """
    # Pre-flight budget check: BUDGET_FULL_POP_HITS=0.0 means "disabled".
    if BUDGET_FULL_POP_HITS <= 0.0:
        logger.info("estimate_full_pop_hits: budget is zero; returning None")
        return None
    start = time.time()
    if blocking_col not in df.columns or df.is_empty():
        return 0
    if (time.time() - start) > BUDGET_FULL_POP_HITS:
        return None
    try:
        counts = (
            df.group_by(blocking_col)
            .agg(pl.len().alias("n"))
            .filter(pl.col("n") > 1)
        )
        if (time.time() - start) > BUDGET_FULL_POP_HITS:
            return None
        n_pairs = int(
            counts.select(
                (pl.col("n") * (pl.col("n") - 1) / 2).sum()
            ).item()
        )
        return n_pairs
    except Exception as exc:
        logger.warning("estimate_full_pop_hits failed: %s", exc)
        return None


def compute_cross_blocking_overlap(
    df: pl.DataFrame, key_a: str, key_b: str,
) -> float | None:
    """Fraction of (record_i, record_j) pairs that are co-blocked under
    BOTH key_a AND key_b out of all pairs co-blocked under EITHER.

    overlap = |co_a INTERSECTION co_b| / |co_a UNION co_b|

    Returns 1.0 if key_a == key_b (degenerate). Returns None on budget.
    Used by rule_cross_blocking_disagreement: low overlap indicates
    blocking key is genuinely capturing wrong candidates (orthogonal
    keys agree on nothing) vs. just-too-few-matches.
    """
    if key_a == key_b:
        return 1.0
    # Pre-flight budget check: BUDGET_CROSS_BLOCKING=0.0 means "disabled".
    if BUDGET_CROSS_BLOCKING <= 0.0:
        logger.info("compute_cross_blocking_overlap: budget is zero; returning None")
        return None
    start = time.time()
    if key_a not in df.columns or key_b not in df.columns or df.is_empty():
        return None
    if (time.time() - start) > BUDGET_CROSS_BLOCKING:
        return None
    try:
        # Co-blocked under key_a: pairs sharing same key_a value
        df_indexed = df.with_row_index("__row__")
        a_pairs = (
            df_indexed.group_by(key_a)
            .agg(pl.col("__row__").alias("rows"))
            .filter(pl.col("rows").list.len() > 1)
        )
        b_pairs = (
            df_indexed.group_by(key_b)
            .agg(pl.col("__row__").alias("rows"))
            .filter(pl.col("rows").list.len() > 1)
        )

        if (time.time() - start) > BUDGET_CROSS_BLOCKING:
            return None

        # Build pair sets (small enough to materialize as Python sets
        # within budget; if too big, return None defensively)
        def _pairs_set(grouped: pl.DataFrame) -> set[tuple[int, int]] | None:
            pairs: set[tuple[int, int]] = set()
            for rows in grouped["rows"].to_list():
                rows = sorted(rows)
                for i in range(len(rows)):
                    for j in range(i + 1, len(rows)):
                        pairs.add((rows[i], rows[j]))
                        if (time.time() - start) > BUDGET_CROSS_BLOCKING:
                            return None
            return pairs

        set_a = _pairs_set(a_pairs)
        if set_a is None:
            return None
        set_b = _pairs_set(b_pairs)
        if set_b is None:
            return None

        union = set_a | set_b
        if not union:
            return 1.0  # no co-blocked pairs at all -> degenerate, treat as match
        intersection = set_a & set_b
        return len(intersection) / len(union)
    except Exception as exc:
        logger.warning("compute_cross_blocking_overlap failed: %s", exc)
        return None


def dispatch_compute_column_priors(df_or_ds):
    """Route to in-memory or distributed compute_column_priors by input type."""
    from goldenmatch.distributed import is_ray_dataset
    if is_ray_dataset(df_or_ds):
        from goldenmatch.distributed.indicators import compute_column_priors_distributed
        return compute_column_priors_distributed(df_or_ds)
    return compute_column_priors(df_or_ds)


def dispatch_estimate_sparse_match_signal(df_or_ds, *, exact_columns):
    from goldenmatch.distributed import is_ray_dataset
    if is_ray_dataset(df_or_ds):
        from goldenmatch.distributed.indicators import estimate_sparse_match_signal_distributed
        return estimate_sparse_match_signal_distributed(df_or_ds, exact_columns=exact_columns)
    return estimate_sparse_match_signal(df_or_ds, exact_columns=exact_columns)


BUDGET_COLLISION = 8.0


def compute_identity_collision_signal(
    df: pl.DataFrame,
    identity_col: str,
    witness_cols: list[str],
) -> CollisionSignal:
    """Detect whether an identity column is shared across distinct entities.

    For each multi-record group (rows sharing the same `identity_col` value),
    compute the max pairwise divergence (1 - similarity) on `witness_cols`.
    Returns the fraction of multi-record groups where max-divergence > 0.5.

    A high rate indicates the identity column is NOT a reliable identity
    anchor (T3's adversarial pattern: same email used for distinct people
    with different addresses, phones, cities).

    Budget: BUDGET_COLLISION seconds. On exhaustion, returns
    CollisionSignal(rate=0.0, witness_used="") sentinel.
    """
    start = time.time()
    if BUDGET_COLLISION <= 0.0:
        return CollisionSignal(rate=0.0, witness_used="")
    if not witness_cols or df.is_empty() or identity_col not in df.columns:
        return CollisionSignal(rate=0.0, witness_used="")
    valid_witnesses = [c for c in witness_cols if c in df.columns]
    if not valid_witnesses:
        return CollisionSignal(rate=0.0, witness_used="")

    try:
        # Group by identity_col; only multi-record groups matter
        groups = (
            df.group_by(identity_col)
            .agg(pl.len().alias("__n__"))
            .filter(pl.col("__n__") > 1)
        )
        if (time.time() - start) > BUDGET_COLLISION:
            return CollisionSignal(rate=0.0, witness_used="")
        if groups.is_empty():
            return CollisionSignal(rate=0.0, witness_used="")

        n_groups = groups.height
        n_high_divergence = 0
        winning_witness = ""
        max_observed_div = 0.0

        # Use rapidfuzz for similarity computation
        from rapidfuzz import fuzz

        for group_value in groups[identity_col].to_list():
            if (time.time() - start) > BUDGET_COLLISION:
                return CollisionSignal(rate=0.0, witness_used="")
            group_df = df.filter(pl.col(identity_col) == group_value)
            n = group_df.height
            if n < 2:
                continue
            max_div_in_group = 0.0
            for witness in valid_witnesses:
                vals = group_df[witness].cast(str).fill_null("").to_list()
                # max pairwise divergence
                for i in range(n):
                    for j in range(i + 1, n):
                        sim = fuzz.token_sort_ratio(vals[i], vals[j]) / 100.0
                        div = 1.0 - sim
                        if div > max_div_in_group:
                            max_div_in_group = div
                            if div > max_observed_div:
                                max_observed_div = div
                                winning_witness = witness
            if max_div_in_group > 0.5:
                n_high_divergence += 1

        rate = n_high_divergence / n_groups if n_groups > 0 else 0.0
        return CollisionSignal(rate=rate, witness_used=winning_witness)
    except Exception as exc:
        logger.warning("compute_identity_collision_signal failed: %s", exc)
        return CollisionSignal(rate=0.0, witness_used="")
